### ISSUE TITLE
updateDPT inconsistent

### ISSUE THREAD

**Issue by Shane32** [joncampbell123/dosbox-x#615]
After mounting an image, the updateDPT() function (within bios_disk.cpp) is called, apparently to update the BIOS disk parameter table.
- Why would it be different for a secondary drive than for the primary drive?
- What about when a drive is released -- shouldn't it clear the variables?

I'd like to extend the BIOS to support 4 hard drives, if this is at all easy to do.  Just accepting 82h and 83h to int13h calls lets MS-DOS 5.0's FDISK command recognize all 4 drives.  (Already prep'd -- just change the MAX_HDD_IMAGES and MAX_DISK_IMAGES constants)
- Is there any code here we need to add for a 3rd and 4th drive?

**Comment by Shane32** [comment]
A cross reference on diskparm0 reveals this interesting snippet from bios_disk.cpp:922:
```
    /* FIXME: Um... these aren't callbacks. Why are they allocated as callbacks? We have ROM general allocation now. */
	diskparm0 = CALLBACK_Allocate();
	CALLBACK_SetDescription(diskparm0,"BIOS Disk 0 parameter table");
	diskparm1 = CALLBACK_Allocate();
	CALLBACK_SetDescription(diskparm1,"BIOS Disk 1 parameter table");
```

**Comment by joncampbell123** [comment]
DOSBox-X already supports more than 2 hard drives.

At least, I have tested that so far with IDE emulation. Not sure about INT 13h, but it should work.

The UpdateDPT function updates two legacy tables stored in the BIOS data area regarding hard drives.

As I understand it, older MFM/RLL hard drives (before IDE) required a fixed disk BIOS to know the geometry, and to store it somewhere where DOS applications and INT 13h could find it.

**Comment by joncampbell123** [comment]
I made that comment in bios_disk.cpp because there's no need to use the callback system to allocate disk parameter tables. I added around that time a dynamic ROM allocation function for allocating regions of the ROM BIOS area, so the DPT tables could easily be allocated from the ROM BIOS.

**Comment by joncampbell123** [comment]
I don't believe the DPT legacy tables support anything beyond two hard drives, so beyond 2 hard drives, don't worry about the DPT.

**Comment by joncampbell123** [comment]
Here's the reason for those DPT tables:

http://www.ctyme.com/intr/int-41.htm

http://www.ctyme.com/intr/int-46.htm

**Comment by Shane32** [comment]
Ok well I must be missing something.  I thought the IDE emulation constructor just has a index pointer to the imageDiskList array - which is coded with only 2 floppies and 2 hard drives.  Once two hard drives are in the imageDiskList array, attempting to imgmount a 3rd hard drive skips trying to add to the IDE controller because there's no spot to cross reference in the imageListDisk array.  Or did i miss something?

**Comment by Shane32** [comment]
If it's ok with you i'll post a pull request that updates the DPT tables with zeros when a drive is unloaded, and sets the max int 13h hard drives to 4 (total 6), which seems to work for DOS anyway

**Comment by joncampbell123** [comment]
I thought imageListDisk[] was for floppy disks only.

**Comment by Shane32** [comment]
the swapList is for floppies only, where you swap floppies out with Ctrl-F4

**Comment by Shane32** [comment]
i'm not sure how to test IDE functionality.  Spinrite doesn't seem to see it

**Comment by joncampbell123** [comment]
It depends on whether IDE emulation is enabled.

By default, the first two IDE controllers should be enabled by default.

There are up to 8 IDE controllers available. Beyond the first two, the extra interfaces need either the resources specified or to have ISA PnP emulation enabled.

**Comment by Shane32** [comment]
if you can tell me how to test IDE emulation, I can play with it and be sure it's working or fix it if not.

**Comment by Shane32** [comment]
Win 95 ?

**Comment by joncampbell123** [comment]
I spoke wrong, IDE emulation refers to imageListDisk[] which in the DOS shell refers to drive letters.

The IDE emulation attaches to those as well.

MAX_DISK_IMAGES is 4, which limits the hard drive images to 2. Try increasing that.

You can verify the IDE I/O ports using the DEBUG command, using the 'i' command to read I/O ports.

Another way is to use the IDE test program in my DOSLIB project:

https://github.com/joncampbell123/doslib

And of course, you can try running Windows 95 with it.

When you run IDE emulation against Windows 95, make sure the "fake v86 io" option is enabled, or Windows 95 won't see the primary IDE controller. It's a hack related to how Windows 95 autodetects which IDE controller the boot drive is associated with.

**Comment by joncampbell123** [comment]
Sorry if I seem slightly inconsistent, I haven't touch the IDE code since working on it 2-8 years ago.

**Comment by Shane32** [comment]
Ok thanks!  I'll play with it some more.

**Comment by Shane32** [comment]
Your insight is really helpful!

**Comment by joncampbell123** [comment]
DOSLIB IDE "toy" program:

https://github.com/joncampbell123/doslib/tree/master/hw/ide

It's easiest to grab a binary build from DOSLIB and obtain it from hw/ide/dos86l or hw/ide/dos386f and run it directly.

**Comment by Shane32** [comment]
Just FYI, mounted hard drive images are 'pushed' on the imageDiskList array regardless of which drive letter they are mounted at.  so if you mount D first, it gets mounted at index 2 (which is int 13h 80h), and then if you mount C next, it gets mounted at index 3 (int 13h 81h).  I'm seeing some stuff online that says that 'modern' BIOSes do it this way, where as some older BIOSes will always mount the controller 1 master at 80h, and the controller 1 slave at 81h, and so on.  Not sure it makes any difference, but another thought would be that we could change the imgmount code so C: always mounts at int 13h 80h and ide 1 master --- and d: always mounts at int 13h 81h and ide 1 slave -- and so on through F:.  It really only matters if you're going to boot from an image, but then, you're probably going to use the imgmount override that sets exactly where you want it mounted at.  So I'd probably leave it as-is.

**Comment by Shane32** [comment]
Based on the information at http://www.ctyme.com/intr/int-41.htm, it would seem that there is a defined way to store 4 hard drives' information in the bios disk parameter table.  I'll need to look at that some more.
